File: ios/TacticalMaps/Calibration/MapSource.swift
import Foundation
import CoreLocation
import MapKit

/// Abstract source of underlying basemap imagery.
///
/// Concrete kinds today:
/// - OnlineRasterBasemapSource - Esri Satellite/Topo, OSM Topo/Street tiles.
/// - PDFMapSource (.geoPDF)  - GeoPDF with neat-line/projection tags parsed.
/// - PDFMapSource (.calibrated) - regular PDF fitted with 3+ fiduciaries.
/// - OfflineTileMapSource - sideloaded MBTiles raster.
///
/// All sources expose a common contract: given WGS84 coord render the correct
/// pixels; given screen point return the underlying WGS84 coord. Overlays
/// (waypoints, drawings) stored in WGS84 and travel between sources unchanged.
protocol MapSource: AnyObject {
    var id: UUID { get }
    var displayName: String { get }
    var kind: MapSourceKind { get }

    /// Region the source can show. nil for satellite (unbounded).
    var coverage: MKCoordinateRegion? { get }

    /// Manual calibration state. GeoPDF placement is carried by the parsed
    /// bounds/placement affine; hand-calibrated PDFs retain their fiduciaries.
    var calibration: Calibration? { get }
}

enum MapSourceKind: String, Codable { case onlineRaster, geoPDF, calibratedPDF, offlineTiles }

/// Calibration metadata for a PDF source.
enum Calibration {
    /// User placed N>=3 fiduciaries, we fit a best-effort affine transform.
    case fiduciaries([Fiduciary], transform: AffineTransform2D)
}

File: android/app/src/main/java/com/tacticalmaps/calibration/MapSource.kt
package com.tacticalmaps.calibration

import com.google.android.gms.maps.model.LatLng
import com.google.android.gms.maps.model.LatLngBounds
import java.util.UUID

/**
 * Abstract basemap source. Three flavours today:
 *  - [AppleSatelliteMapSourceAndroid] — Google satellite, used when no PDF is loaded
 *    (we keep the iOS naming so the cross-platform doc reads consistently).
 *  - [PdfMapSource] in `.geoPDF` mode — calibration parsed from GeoPDF tags.
 *  - [PdfMapSource] in `.calibratedPdf` mode — user-fitted via 3+ fiduciaries.
 *
 * Overlays are stored in WGS84 and travel between sources unchanged.
 */
sealed interface MapSource {
    val id: String
    val displayName: String
    val kind: MapSourceKind
    val coverage: LatLngBounds?
    val calibration: Calibration?
}

enum class MapSourceKind { APPLE_SATELLITE, GEO_PDF, CALIBRATED_PDF }

/** Calibration state for a PDF source. */
sealed interface Calibration {
    data class Parsed(val crs: String, val transform: AffineTransform2D) : Calibration
    data class Fiduciaries(val fids: List<Fiduciary>, val transform: AffineTransform2D) : Calibration
}

class AppleSatelliteMapSourceAndroid : MapSource {
    override val id: String = UUID.randomUUID().toString()
    override val displayName = "Google Satellite"
    override val kind = MapSourceKind.APPLE_SATELLITE
    override val coverage: LatLngBounds? = null
    override val calibration: Calibration? = null
}
